include/deal.II/matrix_free/matrix_free.h

```diff
@@ -802,6 +802,138 @@ class MatrixFree : public Subscriptor
             const InVector &src,
             const bool      zero_dst_vector = false) const;
 
+  /**
+   * This function is similar to the cell_loop with an std::function object to
+   * specify to operation to be performed on cells, but adds two additional
+   * functors to execute some additional work before and after the cell
+   * integrals are computed.
+   *
+   * The two additional functors work on a range of degrees of freedom,
+   * expressed in terms of the degree-of-freedom numbering of the selected
+   * DoFHandler `dof_handler_index_pre_post` in MPI-local indices. The
+   * arguments to the functors represent a range of degrees of freedom at a
+   * granularity of
+   * internal::MatrixFreeFunctions::DoFInfo::chunk_size_zero_vector entries
+   * (except for the last chunk which is set to the number of locally owned
+   * entries) in the form `[first, last)`. The idea of these functors is to
+   * bring operations on vectors closer to the point where they accessed in a
+   * matrix-free loop, with the goal to increase cache hits by temporal
+   * locality. This loop guarantees that the `operation_before_loop` hits all
+   * relevant unknowns before they are first touched in the cell_operation
+   * (including the MPI data exchange), allowing to execute some vector update
+   * that the `src` vector depends upon. The `operation_after_loop` is similar
+   * - it starts to execute on a range of DoFs once all DoFs in that range
+   * have been touched for the last time time by the `cell_operation`
+   * (including the MPI data exchange), allowing e.g. to compute some vector
+   * operations that depend on the result of the current cell loop in `dst` or
+   * want to modify `src`. The efficiency of caching depends on the numbering
+   * of the degrees of freedom because of the granularity of the ranges.
+   *
+   * @param cell_operation Pointer to member function of `CLASS` with the
+   * signature <tt>cell_operation (const MatrixFree<dim,Number> &, OutVector &,
+   * InVector &, std::pair<unsigned int,unsigned int> &)</tt> where the first
+   * argument passes the data of the calling class and the last argument
+   * defines the range of cells which should be worked on (typically more than
+   * one cell should be worked on in order to reduce overheads).
+   *
+   * @param owning_class The object which provides the `cell_operation`
+   * call. To be compatible with this interface, the class must allow to call
+   * `owning_class->cell_operation(...)`.
+   *
+   * @param dst Destination vector holding the result. If the vector is of
+   * type LinearAlgebra::distributed::Vector (or composite objects thereof
+   * such as LinearAlgebra::distributed::BlockVector), the loop calls
+   * LinearAlgebra::distributed::Vector::compress() at the end of the call
+   * internally.
```

Comment:
So if I use a Epetra vector, I have to call `compress()` myself? Why is there a difference with `LA::d::Vector`?

Reply:
The infrastructure does currently not support external vectors anyway because the indices are local in matrix-free, whereas we only have access to global indices in Epetra or PETSc vectors. I will add a comment that those vectors are not supported in parallel, so it also makes no sense to call a method on them.